### PR TITLE
[Merged by Bors] - feat: port Order.Hom.Lattice

### PR DESCRIPTION
**Pull request by mpenciak**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
Zulip discussions of the issues beginning on line 188 are [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Order.2EHom.2ELattice/near/323360873) and [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/What.20is.20an.20outParam.3F/near/323529731). This is all related to https://github.com/leanprover/lean4/issues/1901 . I'm trying to make some progress.

**Comment by kbuzzard** [comment]
There are still plenty of problems with this file (and it doesn't lint either) but I've made a little progress.

**Comment by kbuzzard** [comment]
The dangerous instance lint near the top can apparently be ignored: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/unassigned.20metavariable.20in.20out-param/near/323590485

File: Mathlib/Order/Hom/Lattice.lean [hunk 1]
```diff
@@ -1304,38 +1304,41 @@ end BoundedLatticeHom
 /-! ### Dual homs -/
 
 
+-- Porting note: added `toFun` in everywhere below
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/Hom/Lattice.lean:1295]: All the `toFun`s below avoid timeouts, but are a pretty disappointing fix. @Vierkantor , do you have any thoughts?

Comment (Vierkantor) [reply Mathlib/Order/Hom/Lattice.lean:1295]: The `toFun`s are definitely not the right way to go given all the `simpNF` lints. Maybe replacing the commented-out `CoeFun` instances with `FunLike` will help?

Comment (Vierkantor) [reply Mathlib/Order/Hom/Lattice.lean:1295]: I'm trying to debug this but Lean just times out when I'm looking at the failing `synthInstance` trace. But it seems that the problem lies in not being able to `CoeFun` an `Equiv`, not `SupHom`/etc. Let me think for a bit longer...

Comment (Vierkantor) [reply Mathlib/Order/Hom/Lattice.lean:1295]: I finally got an instance trace to work and :grimacing: it's not going well

```
    [Meta.synthInstance] new goal LinearOrder
          (WithTop
                (WithTop
                    (WithTop
                        (WithTop
                                (WithTop
                                  ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈᵒᵈ
      [Meta.synthInstance.instances] #[@LinearOrderedAddCommMonoid.toLinearOrder, @LinearOrderedCommMonoid.toLinearOrder, @CanonicallyLinearOrderedAddMonoid.toLinearOrder, @CanonicallyLinearOrderedMonoid.toLinearOrder, @LinearOrderedAddCommGroup.toLinearOrder, @LinearOrderedCommGroup.toLinearOrder, @LinearOrderedRing.toLinearOrder, @CompleteLinearOrder.toLinearOrder, OrderDual.linearOrder]
  [Meta.synthInstance.resume] propagating AddMonoid
        (WithTop
            (WithTop
                (WithTop
                    (WithTop
                            (WithTop
                              ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ to subgoal AddMonoid
        (WithTop
            (WithTop
                (WithTop
                    (WithTop
                            (WithTop
                              ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ of AddMonoid
        (WithTop
          (WithTop
              (WithTop
                  (WithTop
                      (WithTop
                              (WithTop
                                ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ)
    [Meta.synthInstance.resume] size: 127
  [Meta.synthInstance.resume] propagating AddMonoid
        (WithTop
          (WithTop
              (WithTop
                  (WithTop
                      (WithTop
                              (WithTop
                                ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ) to subgoal AddMonoid
        (WithTop
          (WithTop
              (WithTop
                  (WithTop
                      (WithTop
                              (WithTop
                                ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ) of AddMonoid
        (WithTop
            (WithTop
                (WithTop
                    (WithTop
                        (WithTop
                                (WithTop
                                  ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ
    [Meta.synthInstance.resume] size: 128
  [Meta.synthInstance.resume] propagating AddMonoid
        (WithTop
          (WithTop
              (WithTop
                  (WithTop
                      (WithTop
                              (WithTop
                                ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ) to subgoal AddMonoid
        (WithTop
          (WithTop
              (WithTop
                  (WithTop
                      (WithTop
                              (WithTop
                                ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ) of AddMonoid
        (WithTop
          (WithTop
            (WithTop
                (WithTop
                    (WithTop
                        (WithTop
                                (WithTop
                                  ℕᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈᵒᵈ))ᵒᵈᵒᵈᵒᵈ)ᵒᵈ)ᵒᵈ)ᵒᵈ))
```

Comment (Vierkantor) [reply Mathlib/Order/Hom/Lattice.lean:1295]: I finally figured it out: adding `FunLike` instances would actually help, but we only want to add it for `Embedding`.

**Comment by Vierkantor** [comment]
All errors should be fixed!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**